apps/aecore/src/aec_paying_for_tx.erl

```diff
@@ -0,0 +1,213 @@
+signers(#paying_for_tx{} = Tx, _) ->
+    {ok, [payer_pubkey(Tx)]}.
+
+-spec process(tx(), aec_trees:trees(), aetx_env:env()) -> {ok, aec_trees:trees(), aetx_env:env()}.
+process(#paying_for_tx{tx = STx} = Tx, Trees, Env0) ->
+    case check_tx(STx) of
+        ok ->
+            PInstr = aeprimop:paying_for_tx_instructions(payer_pubkey(Tx), nonce(Tx), fee(Tx)),
+            case aeprimop:eval(PInstr, Trees, Env0) of
+                {ok, Trees1, Env1} ->
+                    Env2 = aetx_env:set_payer(Env1, payer_pubkey(Tx)),
+                    case aetx_sign:verify_w_env(STx, Trees1, Env2) of
+                        ok ->
+                            aetx:process(aetx_sign:tx(STx), Trees1, Env2);
+                        Err = {error, _} ->
+                            Err
+                    end;
+                Err = {error, _} ->
+                    Err
+            end;
+        Err = {error, _} ->
+            Err
+    end.
+
+serialize(#paying_for_tx{payer_id = PayerId,
+                         nonce    = Nonce,
+                         fee      = Fee,
+                         tx       = InnerTx} = Tx) ->
+    SerTx = aetx_sign:serialize_to_binary(InnerTx),
+    {version(Tx),
+     [ {payer_id, PayerId}
+     , {nonce, Nonce}
+     , {fee, Fee}
+     , {tx, SerTx}
+     ]}.
+
+deserialize(?PAYING_FOR_TX_VSN,
+            [ {payer_id, PayerId}
+            , {nonce, Nonce}
+            , {fee, Fee}
+            , {tx, SerTx}]) ->
+    account = aeser_id:specialize_type(PayerId),
+    InnerTx = aetx_sign:deserialize_from_binary(SerTx),
+    ok      = check_tx(InnerTx),
+    #paying_for_tx{payer_id = PayerId,
+                   nonce    = Nonce,
+                   fee      = Fee,
+                   tx       = InnerTx}.
+
+serialization_template(?PAYING_FOR_TX_VSN) ->
+    [ {payer_id, id}
+    , {nonce, int}
+    , {fee, int}
+    , {tx, binary}
+    ].
+
+for_client(#paying_for_tx{ payer_id = PayerId,
+                           nonce    = Nonce,
+                           fee      = Fee,
+                           tx       = InnerTx}) ->
+    #{<<"payer_id">>  => aeser_api_encoder:encode(id_hash, PayerId),
+      <<"nonce">>     => Nonce,
+      <<"fee">>       => Fee,
+      <<"tx">>        => aetx_sign:serialize_for_client_inner(InnerTx, #{})}.
+
+-spec version(tx()) -> non_neg_integer().
+version(_) ->
+    ?PAYING_FOR_TX_VSN.
+
+-spec valid_at_protocol(aec_hard_forks:protocol_vsn(), tx()) -> boolean().
+valid_at_protocol(P, #paying_for_tx{ tx = STx }) ->
+    P >= ?IRIS_PROTOCOL_VSN andalso
+        aetx:valid_at_protocol(P, aetx_sign:tx(STx)).
+
+%%%===================================================================
+%%% Internal functions
+
+check_tx(STx) ->
+    Tx = aetx_sign:tx(STx),
+    {Type, InnerTx} = aetx:specialize_type(Tx),
+    case Type of
+        paying_for_tx                   -> {error, not_allowed_to_pay_for_recursively};
+        ga_meta_tx                      -> check_tx(aega_meta_tx:tx(InnerTx));
```

Comment:
May the ga_meta_tx in its turn contain a paying_for_tx ? Then there is some indirect recursion.

Reply:
No, that is exactly what we are checking here...